src/DataStreams/RemoteQueryExecutor.cpp

```diff
@@ -100,7 +102,7 @@ RemoteQueryExecutor::~RemoteQueryExecutor()
       * these connections did not remain hanging in the out-of-sync state.
       */
     if (established || isQueryPending())
-        connections->disconnect();
+        ConnectionCollector::enqueueConnectionCleanup(std::move(connections), sync_draining);
```

Comment:
Previously, in case of exception or something, we disconnect.
Now it looks like we will try to drain anyway.
Actually, I do not know if it is a good thing to do. E.g. what if we get an exception from `sendExternalTables` internals, while sending data. If we read from such connection, we could probably wait for the timeout. (Maybe we will not, cause we send cancel inside `enqueueConnectionCleanup`).

Anyway, I think it will be better to remember connection state: explicitly specify if we want to drain it in every place